qa/workunits/rgw/test_rgw_versioning.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
+
+import logging as log
+import json
+import uuid
+import botocore
+from common import exec_cmd, create_user, boto_connect
+from botocore.config import Config
+
+"""
+Tests behavior of bucket versioning.
+"""
+# The test cases in this file have been annotated for inventory.
+# To extract the inventory (in csv format) use the command:
+#
+#   grep '^ *# TESTCASE' | sed 's/^ *# TESTCASE //'
+#
+#
+
+""" Constants """
+USER = 'versioning-tester'
+DISPLAY_NAME = 'Versioning Testing'
+ACCESS_KEY = 'LTA662PVVDTDWX6M2AB0'
+SECRET_KEY = 'pvtchqajgzqx5581t6qbddbkj0bgf3a69qdkjcea'
+BUCKET_NAME = 'versioning-bucket'
+DATA_POOL = 'default.rgw.buckets.data'
+
+def main():
+    """
+    execute versioning tests
+    """
+    create_user(USER, DISPLAY_NAME, ACCESS_KEY, SECRET_KEY)
+
+    connection = boto_connect(ACCESS_KEY, SECRET_KEY, Config(retries = {
+        'total_max_attempts': 1,
+    }))
+
+    # pre-test cleanup
+    try:
+        bucket = connection.Bucket(BUCKET_NAME)
+        bucket.objects.all().delete()
+        bucket.object_versions.all().delete()
+        bucket.delete()
+    except botocore.exceptions.ClientError as e:
+        if not e.response['Error']['Code'] == 'NoSuchBucket':
+            raise
+
+    bucket = connection.create_bucket(Bucket=BUCKET_NAME)
+    connection.BucketVersioning(BUCKET_NAME).enable()
+
+    # reproducer for bug from https://tracker.ceph.com/issues/59663
+    # TESTCASE 'verify that index entries and OLH objects are cleaned up after redundant deletes'
+    log.debug('TEST: verify that index entries and OLH objects are cleaned up after redundant deletes\n')
+    key = str(uuid.uuid4())
+    resp = bucket.Object(key).delete()
+    assert 'DeleteMarker' in resp, 'DeleteMarker key not present in response'
+    assert resp['DeleteMarker'], 'DeleteMarker value not True in response'
+    assert 'VersionId' in resp, 'VersionId key not present in response'
+    version_id = resp['VersionId']
+    bucket.Object(key).delete()
+    connection.ObjectVersion(bucket.name, key, version_id).delete()
+    # bucket index should now be empty
+    out = exec_cmd(f'radosgw-admin bi list --bucket {BUCKET_NAME}')
+    json_out = json.loads(out)
+    assert len(json_out) == 0, 'bucket index was not empty after all objects were deleted'
+
+    (_out, ret) = exec_cmd(f'rados -p {DATA_POOL} ls | grep {key}', check_retcode=False)
+    assert ret != 0, 'olh object was not cleaned up'
+
+    # TESTCASE 'verify that index entries and OLH objects are cleaned up after index linking error'
+    log.debug('TEST: verify that index entries and OLH objects are cleaned up after index linking error\n')
+    key = str(uuid.uuid4())
+    try:
+        exec_cmd('ceph config set client.rgw rgw_debug_inject_set_olh_err 2')
```

Comment:
Wondering if I need a brief sleep between this command and the next one. Seems like it's possible that this config update just wasn't applied prior to the following delete call when run in teuthology.

Reply:
using the running radosgw's admin socket to change config would probably be more reliable, but i'm okay with the sleeps if they pass consistently

Reply:
I'm questioning whether it's really related to the config being applied now, but I'll update the tests again to use the admin socket and will remove the special chars that cause json decoding to fail so that the we fail on the following assertion instead. Any other recommendations for how to debug these failures since they aren't happening with a vstart cluster @cbodley ?

Reply:
the use of `client.rgw` might be problematic. vstart uses entity names like `client.rgw.8000`, but in these teuthology jobs the name is `client.0`

Reply:
to work in both environments, i think we need a command-line option or environment variable to pass in the right entity name or admin socket path

Reply:
i scheduled a job in https://pulpito.ceph.com/cbodley-2023-07-13_17:57:16-rgw:verify-main-distro-default-smithi/ that replaces `client.rgw` with `client.0`

Reply:
that didn't pick up the right suite-branch, rerun in https://pulpito.ceph.com/cbodley-2023-07-13_18:29:03-rgw:verify-main-distro-default-smithi/

Reply:
is there an easy way to get the name of the rgw client that works in both contexts or should i just set the config at the `client` or `global` level instead?

Reply:
http://qa-proxy.ceph.com/teuthology/cbodley-2023-07-13_18:29:03-rgw:verify-main-distro-default-smithi/7336231/teuthology.log passed with `config set client.0`

> should i just set the config at the client or global level instead?

i would expect `client` to work in all cases, yeah

Reply:
updated